File: conversions/leven/elide-leven.rb
#!/usr/bin/env ruby

# Ruby Integration Example for elide-leven
#
# This demonstrates calling the TypeScript Levenshtein distance implementation
# from Ruby using Elide's polyglot capabilities.
#
# Benefits:
# - One string distance implementation shared across TypeScript and Ruby
# - Consistent fuzzy matching across microservices
# - No Ruby Levenshtein gem needed
# - Perfect for Rails search, autocomplete, spell checking

puts "=== Ruby Calling TypeScript Levenshtein Distance ===\n"

# NOTE: Exact syntax depends on Elide's Ruby polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Ruby support is ready

# Assuming Elide provides something like:
# leven_module = Elide.require('./elide-leven.ts')

# Example 1: Basic String Distance
# distance = leven_module.default('cat', 'hat')
# puts "Distance between 'cat' and 'hat': #{distance}"
# puts ""

# Example 2: Rails Search with Fuzzy Matching
# class ProductsController < ApplicationController
#   def search
#     query = params[:q].downcase
#     products = Product.all
#
#     # Find products with fuzzy matching
#     results = products.map do |product|
#       distance = leven_module.default(query, product.name.downcase)
#       [product, distance]
#     end.select { |_, distance| distance <= 3 }
#        .sort_by { |_, distance| distance }
#
#     @products = results.map { |product, _| product }
#   end
# end

# Example 3: Autocomplete Suggestions
# class AutocompleteService
#   def initialize(dictionary)
#     @dictionary = dictionary
#   end
#
#   def suggest(input, max_suggestions: 5)
#     suggestions = @dictionary.map do |word|
#       distance = leven_module.default(input.downcase, word.downcase)
#       [word, distance]
#     end.select { |_, distance| distance <= 3 }
#        .sort_by { |_, distance| distance }
#        .take(max_suggestions)
#        .map { |word, _| word }
#
#     suggestions
#   end
# end
#
# autocomplete = AutocompleteService.new(['ruby', 'rails', 'sinatra', 'hanami'])
# puts "Autocomplete for 'rby':"
# puts autocomplete.suggest('rby').inspect
# puts ""

# Example 4: Sidekiq Job for Data Deduplication
# class DeduplicateCustomersJob
#   include Sidekiq::Worker
#
#   def perform(customer_ids)
#     customers = Customer.where(id: customer_ids)
#     duplicates = []
#
#     customers.each_with_index do |customer1, i|
#       customers[i+1..-1].each do |customer2|
#         distance = leven_module.default(
#           customer1.name.downcase,
#           customer2.name.downcase
#         )
#
#         if distance <= 2
#           duplicates << [customer1.id, customer2.id, distance]
#         end
#       end
#     end
#
#     # Create duplicate alerts
#     duplicates.each do |id1, id2, distance|
#       DuplicateAlert.create!(
#         customer_id_1: id1,
#         customer_id_2: id2,
#         similarity_score: distance
#       )
#     end
#
#     puts "Found #{duplicates.length} potential duplicates"
#   end
# end

# Example 5: Command Suggestion for Rake Tasks
# def suggest_rake_task(input, available_tasks)
#   suggestion = leven_module.closestMatch(
#     input,
#     available_tasks,
#     { maxDistance: 3 }
#   )
#   suggestion
# end
#
# rake_tasks = ['db:migrate', 'db:seed', 'test', 'assets:precompile']
# user_input = 'db:migarte'
# suggestion = suggest_rake_task(user_input, rake_tasks)
# if suggestion
#   puts "Task '#{user_input}' not found."
#   puts "Did you mean '#{suggestion}'?"
# end
# puts ""

puts "Real-world use cases:"
puts "- Rails fuzzy search for products/users"
puts "- Autocomplete APIs with typo tolerance"
puts "- Data deduplication in background jobs"
puts "- Command-line tool suggestions"
puts "- Natural language query processing"
puts ""

puts "Example: Rails E-Commerce Platform"
puts "┌─────────────────────────────────────┐"
puts "│   Elide Leven (TypeScript)         │"
puts "│   conversions/leven/elide-leven.ts │"
puts "└─────────────────────────────────────┘"
puts "         ↓           ↓           ↓"
puts "    ┌────────┐  ┌────────┐  ┌────────┐"
puts "    │ Node.js│  │  Rails │  │ Sidekiq│"
puts "    │  API   │  │  App   │  │ Worker │"
puts "    └────────┘  └────────┘  └────────┘"
puts "         ↓           ↓           ↓"
puts "    Same fuzzy matching everywhere!"
puts ""

puts "Problem Solved:"
puts "Before: Ruby Levenshtein gem + JS leven = different results"
puts "After: One Elide implementation = 100% consistent distances"
puts ""

puts "Rails Integration (when ready):"
puts "  # config/initializers/leven.rb"
puts "  LEVEN = Elide.require('./elide-leven.ts')"
puts "  "
puts "  # app/controllers/search_controller.rb"
puts "  class SearchController < ApplicationController"
puts "    def fuzzy_search"
puts "      distance = LEVEN.default(params[:q], product.name)"
puts "      # Process search with consistent distance"
puts "    end"
puts "  end"
puts ""

puts "Sidekiq Integration (when ready):"
puts "  class FuzzyMatchJob"
puts "    include Sidekiq::Worker"
puts "    "
puts "    def perform(str1, str2)"
puts "      distance = LEVEN.default(str1, str2)"
puts "      # Use consistent Levenshtein distance"
puts "    end"
puts "  end"

File: conversions/leven/elide-leven.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-leven

This demonstrates calling the TypeScript Levenshtein distance implementation
from Python using Elide's polyglot capabilities.

Benefits:
- One string distance implementation shared across TypeScript and Python
- Consistent fuzzy matching across services
- No Python Levenshtein library needed
- Perfect for spell checkers, fuzzy search, autocomplete
"""

# NOTE: Exact syntax depends on Elide's Python polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Python support is ready

# Assuming Elide provides something like:
# from elide import require
# leven_module = require('./elide-leven.ts')

print("=== Python Calling TypeScript Levenshtein Distance ===\n")

# Example 1: Basic String Distance
# distance = leven_module.default('cat', 'hat')
# print(f"Distance between 'cat' and 'hat': {distance}")
# print()

# Example 2: Spell Checker for User Input
# def spell_check(word, dictionary):
#     """Find closest match in dictionary"""
#     best_match = leven_module.closestMatch(word, dictionary)
#     if best_match:
#         distance = leven_module.default(word, best_match)
#         return best_match, distance
#     return None, None
#
# dictionary = ['python', 'typescript', 'javascript', 'rust', 'golang']
# typo = 'typescrpit'
# suggestion, dist = spell_check(typo, dictionary)
# print(f"User typed: '{typo}'")
# print(f"Did you mean: '{suggestion}' (distance: {dist})")
# print()

# Example 3: Fuzzy Search for Product Names
# class ProductSearch:
#     def __init__(self):
#         self.products = [
#             'iPhone 15 Pro',
#             'Samsung Galaxy S24',
#             'Google Pixel 8',
#             'OnePlus 12',
#             'Xiaomi 14 Ultra'
#         ]
#
#     def search(self, query, max_distance=3):
#         """Search products with fuzzy matching"""
#         results = []
#         for product in self.products:
#             distance = leven_module.default(
#                 query.lower(),
#                 product.lower(),
#                 {'maxDistance': max_distance}
#             )
#             if distance <= max_distance:
#                 results.append((product, distance))
#         return sorted(results, key=lambda x: x[1])
#
# search_engine = ProductSearch()
# query = 'ipone 15'  # Typo in search
# results = search_engine.search(query)
# print(f"Search query: '{query}'")
# print("Results:")
# for product, distance in results:
#     print(f"  - {product} (distance: {distance})")
# print()

# Example 4: Data Deduplication
# def deduplicate_names(names, threshold=2):
#     """Find and group similar names"""
#     seen = set()
#     duplicates = []
#
#     for i, name1 in enumerate(names):
#         if name1 in seen:
#             continue
#         group = [name1]
#         seen.add(name1)
#
#         for name2 in names[i+1:]:
#             if name2 in seen:
#                 continue
#             distance = leven_module.default(name1.lower(), name2.lower())
#             if distance <= threshold:
#                 group.append(name2)
#                 seen.add(name2)
#
#         if len(group) > 1:
#             duplicates.append(group)
#
#     return duplicates
#
# customer_names = [
#     'John Smith',
#     'Jon Smith',
#     'Jane Doe',
#     'Jane Do',
#     'Bob Johnson',
#     'Robert Johnson'
# ]
# groups = deduplicate_names(customer_names)
# print("Potential duplicate customer names:")
# for group in groups:
#     print(f"  {' ≈ '.join(group)}")
# print()

# Example 5: Command-Line Suggestion
# def suggest_command(user_input, available_commands):
#     """Suggest correct command for CLI typos"""
#     suggestion = leven_module.closestMatch(
#         user_input,
#         available_commands,
#         {'maxDistance': 3}
#     )
#     return suggestion
#
# commands = ['install', 'build', 'test', 'deploy', 'run', 'dev']
# user_cmd = 'isntall'  # Common typo
# suggestion = suggest_command(user_cmd, commands)
# if suggestion:
#     print(f"Command '{user_cmd}' not found.")
#     print(f"Did you mean '{suggestion}'?")
# print()

print("Real-world use cases:")
print("- Spell checkers and autocomplete systems")
print("- Fuzzy search in e-commerce catalogs")
print("- Data deduplication and record linkage")
print("- CLI command suggestions")
print("- Natural language processing pipelines")
print()

print("Example: E-Commerce Search")
print("┌─────────────────────────────────────┐")
print("│   Elide Leven (TypeScript)         │")
print("│   conversions/leven/elide-leven.ts │")
print("└─────────────────────────────────────┘")
print("         ↓                   ↓")
print("    ┌────────┐          ┌────────┐")
print("    │ Node.js│          │ Python │")
print("    │  API   │          │ Search │")
print("    └────────┘          └────────┘")
print("         ↓                   ↓")
print("    Same fuzzy matching everywhere!")
print()

print("Problem Solved:")
print("Before: Python Levenshtein + JS leven = different algorithms")
print("After: One Elide implementation = 100% consistent distances")
print()

print("Performance Benefits:")
print("- Zero cold start overhead with Elide")
print("- Instant distance calculations")
print("- Shared runtime across languages")
print("- Optimized algorithm with early termination")
print()

print("When Elide Python API is ready, usage will be:")
print("  from elide import require")
print("  leven = require('./elide-leven.ts')")
print("  distance = leven.default('cat', 'hat')  # That's it!")
